File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseIndexingInsteadOfLinqMethods.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;

static class Program
{
    static void Main() { }

    static void InterfaceCases()
    {
        IList<int> ilist = new List<int>();

        ilist.First(); // Noncompliant
        ilist.Last(); // Noncompliant
        ilist.ElementAt(42); // Noncompliant

        ilist?.First(); // Noncompliant
        ilist?.Last(); // Noncompliant
        ilist?.ElementAt(42); // Noncompliant

        ((IList<int>)new List<int> { 42 }).First(); // Noncompliant
        ((IList<int>)new List<int> { 42 })?.Last(); // Noncompliant
    }

    static void ListCases()
    {
        var list = new List<int>();

        var badFirst = list.First(); // Noncompliant {{Indexing at 0 should be used instead of the "Enumerable" extension method "First"}}
        //                  ^^^^^
        var badLast = list.Last(); // Noncompliant {{Indexing at Count-1 should be used instead of the "Enumerable" extension method "Last"}}
        //                 ^^^^
        var badElementAt = list.ElementAt(42); // Noncompliant {{Indexing should be used instead of the "Enumerable" extension method "ElementAt"}}
        //                      ^^^^^^^^^
        var badFirstNullable = list?.First(); // Noncompliant {{Indexing at 0 should be used instead of the "Enumerable" extension method "First"}}
        //                           ^^^^^
        var badLastNullable = list?.Last(); // Noncompliant {{Indexing at Count-1 should be used instead of the "Enumerable" extension method "Last"}}
        //                          ^^^^
        var badElementAtNullable = list?.ElementAt(42); // Noncompliant {{Indexing should be used instead of the "Enumerable" extension method "ElementAt"}}
        //                               ^^^^^^^^^

        Func<List<int>, int> func = l => l.First(); // Noncompliant

        List<int> DoWork() => null;

        DoWork().First(); // Noncompliant
        DoWork().Last(); // Noncompliant
        DoWork().ElementAt(42); // Noncompliant

        DoWork()?.First(); // Noncompliant
        DoWork()?.Last(); // Noncompliant
        DoWork()?.ElementAt(42); // Noncompliant

        new List<int> { 42 }.First(); // Noncompliant
        new List<int> { 42 }.Last(); // Noncompliant
        new List<int> { 42 }.ElementAt(42); // Noncompliant
    }

    static void IListImplementation()
    {
        var implementsIList = new ImplementsIList<int>();

        (true ? implementsIList : implementsIList).First(); // Noncompliant
        (true ? implementsIList : implementsIList).Last(); // Noncompliant
        (true ? implementsIList : implementsIList).ElementAt(42); // Noncompliant

        (implementsIList ?? implementsIList).First(); // Noncompliant
        (implementsIList ?? implementsIList).Last(); // Noncompliant
        (implementsIList ?? implementsIList).ElementAt(42); // Noncompliant

        (implementsIList ?? (true ? implementsIList : implementsIList)).First(); // Noncompliant
        (implementsIList ?? (true ? implementsIList : implementsIList)).Last(); // Noncompliant
        (implementsIList ?? (true ? implementsIList : implementsIList)).ElementAt(42); // Noncompliant

        implementsIList.Fluent().Fluent().Fluent().Fluent().First(); // Noncompliant
        implementsIList.Fluent().Fluent().Fluent().Fluent()?.Last(); // Noncompliant
        implementsIList.Fluent().Fluent().Fluent()?.Fluent().ElementAt(42); // Noncompliant
        implementsIList.Fluent().Fluent().Fluent()?.Fluent()?.First(); // Noncompliant
        implementsIList.Fluent().Fluent()?.Fluent().Fluent().Last(); // Noncompliant
        implementsIList.Fluent().Fluent()?.Fluent().Fluent()?.ElementAt(42); // Noncompliant
        implementsIList.Fluent().Fluent()?.Fluent()?.Fluent().First(); // Noncompliant
        implementsIList.Fluent().Fluent()?.Fluent()?.Fluent()?.Last(); // Noncompliant
        implementsIList.Fluent()?.Fluent().Fluent().Fluent().ElementAt(42); // Noncompliant
        implementsIList.Fluent()?.Fluent().Fluent().Fluent()?.First(); // Noncompliant
        implementsIList.Fluent()?.Fluent().Fluent()?.Fluent().Last(); // Noncompliant
        implementsIList.Fluent()?.Fluent().Fluent()?.Fluent()?.ElementAt(42); // Noncompliant
        implementsIList.Fluent()?.Fluent()?.Fluent().Fluent().First(); // Noncompliant
        implementsIList.Fluent()?.Fluent()?.Fluent().Fluent()?.Last(); // Noncompliant
        implementsIList.Fluent()?.Fluent()?.Fluent()?.Fluent().ElementAt(42); // Noncompliant
        implementsIList.Fluent()?.Fluent()?.Fluent()?.Fluent()?.First(); // Noncompliant
        //                                                      ^^^^^

        implementsIList.First(x => x == 42); // Compliant, calls with the predicate cannot be replaced with indexes
        implementsIList.Last(x => x == 42); // Compliant, calls with the predicate cannot be replaced with indexes

        void AcceptsFirstOrLast<T>(Func<T> methodWithNoArguments)
        { }
        void AcceptsElementAt<T>(Func<int, T> methodWithOneArgument)
        { }

        AcceptsFirstOrLast(implementsIList.First); //FN this is not an invocation, just a member access
        AcceptsFirstOrLast(implementsIList.Last); //FN this is not an invocation, just a member access
        AcceptsElementAt(implementsIList.ElementAt); //FN this is not an invocation, just a member access
    }

    static void IReadonlyListImplementation()
    {
        var readonlyList = new ImplementsIReadonlyList<int>();
        readonlyList.First(); // Noncompliant
        readonlyList.Last(); // Noncompliant
        readonlyList.ElementAt(42); // Noncompliant

        readonlyList?.First(); // Noncompliant
        readonlyList?.Last(); // Noncompliant
        readonlyList?.ElementAt(42); // Noncompliant

        object obj = null;
        ((IReadOnlyList<int>)obj).First(); // Noncompliant
        ((IReadOnlyList<int>)null).Last(); // Noncompliant
        (null as IReadOnlyList<int>).ElementAt(42); // Noncompliant
    }

    static void TrueNegatives()
    {
        var list = new List<int>();
        _ = list[0]; // Compliant
        _ = list[list.Count - 1]; // Compliant
        _ = list[42]; // Compliant

        T First<T>() => default(T);
        T Last<T>() => default(T);
        T ElementAt<T>(int index) => default(T);

        First<int>(); // Compliant
        Last<int>(); // Compliant
        ElementAt<int>(42); // Compliant

        var fakeList = new FakeList<int>();

        fakeList.First(); // Compliant
        fakeList.Last(); // Compliant
        fakeList.ElementAt(42); // Compliant

        var doesNotImplementIList = new DoesNotImplementIList<int>();

        doesNotImplementIList.First(); // Compliant, does not implement IList
        doesNotImplementIList.Last(); // Compliant, does not implement IList
        doesNotImplementIList.ElementAt(42); // Compliant, does not implement IList
    }
}

class ImplementsIList<T> : IList<T>
{
    public ImplementsIList<T> Fluent() => this;
    public T this[int index] { get => default(T); set => value = default(T); }

    // Everything below is just to satisfy IList<T> implementation, useless otherwise
    public int Count => throw new NotImplementedException();
    public bool IsReadOnly => throw new NotImplementedException();
    public void Add(T item) { }
    public void Clear() { }
    public bool Contains(T item) => false;
    public void CopyTo(T[] array, int arrayIndex) { }
    public int IndexOf(T item) => 42;
    public void Insert(int index, T item) { }
    public bool Remove(T item) => false;
    public void RemoveAt(int index) { }
    public IEnumerator<T> GetEnumerator() => null;
    IEnumerator IEnumerable.GetEnumerator() => null;
}

class ImplementsIReadonlyList<T> : IReadOnlyList<T>
{
    public T this[int index] => default(T);
    public int Count => 42;
    public IEnumerator<T> GetEnumerator() => null;
    IEnumerator IEnumerable.GetEnumerator() => null;
}

class DoesNotImplementIList<T> : IEnumerable<T>
{
    public IEnumerator<T> GetEnumerator() => null;
    IEnumerator IEnumerable.GetEnumerator() => null;
}

class FakeList<T> { }

static class FakeListExtensions
{
    public static TSource First<TSource>(this FakeList<TSource> source) => default(TSource);
    public static TSource Last<TSource>(this FakeList<TSource> source) => default(TSource);
    public static TSource ElementAt<TSource>(this FakeList<TSource> source, int index) => default(TSource);
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/UseIndexingInsteadOfLinqMethods.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Linq

Module Module1
    Sub Main()
        Dim list As New List(Of Integer)()

        Dim dummy = list(0) ' Compliant
        dummy = list(list.Count - 1) ' Compliant
        dummy = list(42) ' Compliant

        Dim badFirst = list.First() ' Noncompliant {{Indexing at 0 should be used instead of the "Enumerable" extension method "First"}}
        '                   ^^^^^
        Dim badLast = list.Last() ' Noncompliant {{Indexing at Count-1 should be used instead of the "Enumerable" extension method "Last"}}
        '                  ^^^^
        Dim badElementAt = list.ElementAt(42) ' Noncompliant {{Indexing should be used instead of the "Enumerable" extension method "ElementAt"}}
        '                       ^^^^^^^^^
        Dim badFirstNullable = list?.First() ' Noncompliant {{Indexing at 0 should be used instead of the "Enumerable" extension method "First"}}
        '                            ^^^^^
        Dim badLastNullable = list?.Last() ' Noncompliant {{Indexing at Count-1 should be used instead of the "Enumerable" extension method "Last"}}
        '                           ^^^^
        Dim badElementAtNullable = list?.ElementAt(42) ' Noncompliant {{Indexing should be used instead of the "Enumerable" extension method "ElementAt"}}
        '                                ^^^^^^^^^

        Dim func As Func(Of List(Of Integer), Integer) = Function(l) l.First() ' Noncompliant

        dummy = DoWork().First() ' Noncompliant
        dummy = DoWork().Last() ' Noncompliant
        dummy = DoWork().ElementAt(42) ' Noncompliant

        dummy = DoWork()?.First() ' Noncompliant
        dummy = DoWork()?.Last() ' Noncompliant
        dummy = DoWork()?.ElementAt(42) ' Noncompliant

        dummy = (New List(Of Integer)() From {42}).First() ' Noncompliant
        dummy = (New List(Of Integer)() From {42}).Last() ' Noncompliant
        dummy = (New List(Of Integer)() From {42}).ElementAt(42) ' Noncompliant

        Dim inlineInitialization = {42}.First() ' FN, .GetTypeInfo(CollectionInitializer) returns null
        inlineInitialization = {42}.Last() ' FN, .GetTypeInfo(CollectionInitializer) returns null
        inlineInitialization = {42}.ElementAt(42) ' FN, .GetTypeInfo(CollectionInitializer) returns null

        Dim implementsIList = New ImplementsIList(Of Integer)()

        implementsIList.Fluent().Fluent().Fluent().Fluent().First() ' Noncompliant
        implementsIList.Fluent().Fluent().Fluent().Fluent()?.Last() ' Noncompliant
        implementsIList.Fluent().Fluent().Fluent()?.Fluent().ElementAt(42) ' Noncompliant
        implementsIList.Fluent().Fluent().Fluent()?.Fluent()?.First() ' Noncompliant
        implementsIList.Fluent().Fluent()?.Fluent().Fluent().Last() ' Noncompliant
        implementsIList.Fluent().Fluent()?.Fluent().Fluent()?.ElementAt(42) ' Noncompliant
        implementsIList.Fluent().Fluent()?.Fluent()?.Fluent().First() ' Noncompliant
        implementsIList.Fluent().Fluent()?.Fluent()?.Fluent()?.Last() ' Noncompliant
        implementsIList.Fluent()?.Fluent().Fluent().Fluent().ElementAt(42) ' Noncompliant
        implementsIList.Fluent()?.Fluent().Fluent().Fluent()?.First() ' Noncompliant
        implementsIList.Fluent()?.Fluent().Fluent()?.Fluent().Last() ' Noncompliant
        implementsIList.Fluent()?.Fluent().Fluent()?.Fluent()?.ElementAt(42) ' Noncompliant
        implementsIList.Fluent()?.Fluent()?.Fluent().Fluent().First() ' Noncompliant
        implementsIList.Fluent()?.Fluent()?.Fluent().Fluent()?.Last() ' Noncompliant
        implementsIList.Fluent()?.Fluent()?.Fluent()?.Fluent().ElementAt(42) ' Noncompliant
        implementsIList.Fluent()?.Fluent()?.Fluent()?.Fluent()?.First() ' Noncompliant
        '                                                       ^^^^^

        implementsIList.First(Function(x) x = 42) ' Compliant, calls with the predicate cannot be replaced with indexes
        implementsIList.Last(Function(x) x = 42) ' Compliant, calls with the predicate cannot be replaced with indexes

        First(Of Integer)() ' Compliant
        Last(Of Integer)() ' Compliant
        ElementAt(Of Integer)(42) ' Compliant

        Dim fakeList As New FakeList(Of Integer)()

        fakeList.First() ' Compliant
        fakeList.Last() ' Compliant
        fakeList.ElementAt(42) ' Compliant

    End Sub

    Function DoWork() As List(Of Integer)
        Return Nothing
    End Function

    Function First(Of T)() As T
        Return Nothing
    End Function

    Function Last(Of T)() As T
        Return Nothing
    End Function

    Function ElementAt(Of T)(index As Integer) As T
        Return Nothing
    End Function

    Class ImplementsIList(Of T)
        Implements IList(Of T)

        Public Function Fluent() As ImplementsIList(Of T)
            Return Me
        End Function

        Default Public Property Item(ByVal index As Integer) As T Implements IList(Of T).Item
            Get
                Return Nothing
            End Get
            Set(ByVal value As T)
                value = Nothing
            End Set
        End Property

        Public ReadOnly Property Count As Integer Implements IList(Of T).Count
            Get
                Return 42
            End Get
        End Property

        Public ReadOnly Property IsReadOnly As Boolean Implements IList(Of T).IsReadOnly
            Get
                Return True
            End Get
        End Property

        Public Sub Add(ByVal item As T) Implements IList(Of T).Add
        End Sub

        Public Sub Clear() Implements IList(Of T).Clear
        End Sub

        Public Function Contains(ByVal item As T) As Boolean Implements IList(Of T).Contains
            Return False
        End Function

        Public Sub CopyTo(ByVal array() As T, ByVal arrayIndex As Integer) Implements IList(Of T).CopyTo
        End Sub

        Public Function IndexOf(ByVal item As T) As Integer Implements IList(Of T).IndexOf
            Return 42
        End Function

        Public Sub Insert(ByVal index As Integer, ByVal item As T) Implements IList(Of T).Insert
        End Sub

        Public Function Remove(ByVal item As T) As Boolean Implements IList(Of T).Remove
            Return False
        End Function

        Public Sub RemoveAt(ByVal index As Integer) Implements IList(Of T).RemoveAt
        End Sub

        Public Function GetEnumerator() As IEnumerator(Of T) Implements IEnumerable(Of T).GetEnumerator
            Return Nothing
        End Function

        Private Function IEnumerable_GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
            Return Nothing
        End Function
    End Class

End Module

Class FakeList(Of T)
End Class

Module FakeListExtensions
    <System.Runtime.CompilerServices.Extension>
    Public Function First(Of TSource)(ByVal source As FakeList(Of TSource)) As TSource
        Return Nothing
    End Function

    <System.Runtime.CompilerServices.Extension>
    Public Function Last(Of TSource)(ByVal source As FakeList(Of TSource)) As TSource
        Return Nothing
    End Function

    <System.Runtime.CompilerServices.Extension>
    Public Function ElementAt(Of TSource)(ByVal source As FakeList(Of TSource), ByVal index As Integer) As TSource
        Return Nothing
    End Function
End Module

